// lib/data-scraper.js
var cheerio = require('cheerio');
var moment = require('moment-timezone');

var DataScraper = module.exports = function DataScraper(pageData) {
  this.$ = cheerio.load(pageData);
};

DataScraper.prototype.getMealLinks = function getMealLinks() {
  var $ = this.$;
  var $mealLinks = $('h1 + table tr a');
  return Array.prototype.map.call($mealLinks, function (elem) {
    return $(elem).attr('href');
  });
};

DataScraper.prototype.getMealInfo = function getMealData() {
  var $ = this.$;
  var $mealElements = $('h1 ~ table').first().find('tr');
  var info = {
    raw: {
      title: $('h1').text()
    }
  };

  $mealElements.each(function () {
    var $title = $(this).find('td').first();
    var $data = $title.next();

    if ($title.text().search("Date") != -1) {
      info.raw.date = $data.text();
      return;
    }

    if ($title.text().search("Start time") != -1) {
      info.raw.startTime = $data.text();
      return;
    }
  });

  info.title = info.raw.title;
  var dateString = (info.raw.date + ' ' + info.raw.startTime).replace(/\s+/g, ' ');

  info.date = moment.tz(dateString, 'dddd-D-MMMM-YYYY-H:mma', 'Europe/London')

  if (!info.date.isValid()) {
    // If there's no date we can infer, there's no point continuing
    return null;
  }

  return info;
};
