// lib/user-mocker.js
var Zombie = require('zombie');
var cheerio = require('cheerio');
var DataScraper = require('lib/data-scraper');

var UserMocker = module.exports = function UserMocker() {
  var self = this;

  this.browser = new Zombie();
  this.scraper = new DataScraper('');

  this.browser.on('loaded', function (document) {
    self.scraper.setPageContent(document.documentElement.innerHTML);
  });

  this.loggedIn = false;
};

UserMocker.prototype.loginFromCredentials = function loginFromCredentials(username, password) {
  var self = this;
  return this.browser.visit(UserMocker.BOOKING_URL)
    .then(function () {
      if (self.scraper.isLoggingIn()) {
        // We're logging in
        return self.browser
          .fill('userid', username)
          .fill('pwd', password)
          .pressButton('Login');
      }
    })
    .then(function () {

      self._loginCheck();

    })
};

UserMocker.prototype.loginFromCookies = function loginFromCookies(cookies) {
  var self = this;

  for (var cookie of cookies) {
    this.browser.setCookie(cookie);
  }

  this.getCookies();

  return this.browser.visit(UserMocker.BOOKING_URL)
    .then(function () {
      self._loginCheck();
    });
}

UserMocker.prototype._loginCheck = function loginCheck() {
  if (this.scraper.isLoggingIn()) {
    throw new Error('Credentials incorrect');
  }

  this.loggedIn = true;
}

UserMocker.prototype.getCookies = function getCookies() {
  return this.browser.cookies.map(function (cookie) {
    cookie.name = cookie.key;
    return cookie;
  });
};

UserMocker.prototype.getMealInfo = function () {

};

UserMocker.BOOKING_URL = 'https://www.mealbookings.cai.cam.ac.uk/';
